File: heejae/baekjoon/예산.js
const input = require("fs").readFileSync(process.platform === 'linux' ? "/dev/stdin" : __dirname + "/input.txt")
s = input.toString().trim().split('\n');
const req = s[1].split(' ').map(Number)
const max = Number(s[2]);
let left = 0;
let right = Math.max(...req)
while (left < right) {
    const mid = Math.ceil((left + right) / 2);
    let sum = req.reduce((a, c) => a + (c > mid ? mid : c),0);
    sum > max ? right = mid - 1 : left = mid;
}
console.log(left)


// let [[N], I, [M]] = `${require('fs').readFileSync(0)}`.trim().split`\n`.map(e => e.split` `.map(Number))
// let [l, r] = [0, Math.max(...I)]
// while (l < r) {
//   const mid = Math.ceil((l + r) / 2)
//   I.reduce((a, b) => a + (b > mid ? mid : b), 0) > M ? (r = mid - 1) : (l = mid)
// }
// console.log(l)

File: minji/이분탐색/예산.py
import sys
n = int(sys.stdin.readline().rstrip())
array = list(map(int, sys.stdin.readline().split()))
m = int(sys.stdin.readline().rstrip())

start = 0
end = max(array)

while start <= end:
    mid = (start + end) // 2
    curr = 0
    for i in array:
        if i >= mid:
            curr += mid
        else:
            curr += i
    if curr <= m:
        start = mid + 1
    else:
        end = mid - 1

print(end)